containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2904 +275,1189 @@
-  foldr = foldr
-  {-# INLINE foldr #-}
-  foldl = foldl
-  {-# INLINE foldl #-}
-  foldMap f t = go t
-    where go Nil = mempty
-          go (Tip _ v) = f v
-          go (Bin _ m l r)
-            | m < 0     = go r `mappend` go l
-            | otherwise = go l `mappend` go r
-  {-# INLINE foldMap #-}
-  foldl' = foldl'
-  {-# INLINE foldl' #-}
-  foldr' = foldr'
-  {-# INLINE foldr' #-}
-#if MIN_VERSION_base(4,8,0)
-  length = size
-  {-# INLINE length #-}
-  null   = null
-  {-# INLINE null #-}
-  toList = elems -- NB: Foldable.toList /= IntMap.toList
-  {-# INLINE toList #-}
-  elem = go
-    where go !_ Nil = False
-          go x (Tip _ y) = x == y
-          go x (Bin _ _ l r) = go x l || go x r
-  {-# INLINABLE elem #-}
-  maximum = start
-    where start Nil = error "Data.Foldable.maximum (for Data.IntMap): empty map"
-          start (Tip _ y) = y
-          start (Bin _ m l r)
-            | m < 0     = go (start r) l
-            | otherwise = go (start l) r
-
-          go !m Nil = m
-          go m (Tip _ y) = max m y
-          go m (Bin _ _ l r) = go (go m l) r
-  {-# INLINABLE maximum #-}
-  minimum = start
-    where start Nil = error "Data.Foldable.minimum (for Data.IntMap): empty map"
-          start (Tip _ y) = y
-          start (Bin _ m l r)
-            | m < 0     = go (start r) l
-            | otherwise = go (start l) r
-
-          go !m Nil = m
-          go m (Tip _ y) = min m y
-          go m (Bin _ _ l r) = go (go m l) r
-  {-# INLINABLE minimum #-}
-  sum = foldl' (+) 0
-  {-# INLINABLE sum #-}
-  product = foldl' (*) 1
-  {-# INLINABLE product #-}
+instance Functor (Node t) where
+    fmap _ Tip = Tip
+    fmap f (Bin k v l r) = Bin k (f v) (fmap f l) (fmap f r)
+
+instance Data.Foldable.Foldable IntMap where
+    foldMap f = start
+      where
+        start (IntMap Empty) = mempty
+        start (IntMap (NonEmpty _ minV root)) = f minV `mappend` goL root
+
+        goL Tip = mempty
+        goL (Bin _ maxV l r) = goL l `mappend` goR r `mappend` f maxV
+
+        goR Tip = mempty
+        goR (Bin _ minV l r) = f minV `mappend` goL l `mappend` goR r
+
+    foldr = foldr
+    foldl = foldl
+#if MIN_VERSION_base(4,6,0)
+    foldr' = foldr'
+    foldl' = foldl'
 #endif
 
--- | Traverses in order of increasing key.
 instance Traversable IntMap where
-    traverse f = traverseWithKey (\_ -> f)
-    {-# INLINE traverse #-}
+    traverse f = start
+      where
+        start (IntMap Empty) = pure (IntMap Empty)
+        start (IntMap (NonEmpty min minV node)) = (\minV' root' -> IntMap (NonEmpty min minV' root')) <$> f minV <*> goL node
 
-instance NFData a => NFData (IntMap a) where
-    rnf Nil = ()
-    rnf (Tip _ v) = rnf v
-    rnf (Bin _ _ l r) = rnf l `seq` rnf r
+        goL Tip = pure Tip
+        goL (Bin max maxV l r) = (\l' r' v' -> Bin max v' l' r') <$> goL l <*> goR r <*> f maxV
+
+        goR Tip = pure Tip
+        goR (Bin min minV l r) = Bin min <$> f minV <*> goL l <*> goR r
```

Comment:
Use `liftA2 f x y` instead of `f <$> x <*> y`, `liftA2 f x y <*> z` instead of `f <$> x <*> y <*> z`, etc. This could matter if the functor has a custom `liftA2`.